store/store.go

```diff
@@ -244,82 +249,187 @@ func (bs *BlockStore) LoadSeenCommit(height int64) *types.Commit {
-		// We can't trust batches to be atomic, so update base first to make sure noone
-		// tries to access missing blocks.
-		bs.mtx.Lock()
-		bs.base = base
-		bs.mtx.Unlock()
-		bs.saveState()
-
-		err := batch.WriteSync()
-		if err != nil {
-			return fmt.Errorf("failed to prune up to height %v: %w", base, err)
-		}
-		batch.Close()
-		return nil
-	}
 
-	for h := base; h < height; h++ {
-		meta := bs.LoadBlockMeta(h)
-		if meta == nil { // assume already deleted
-			continue
-		}
-		if err := batch.Delete(calcBlockMetaKey(h)); err != nil {
-			return 0, err
+	pruned := uint64(0)
+	for iter.Valid() {
+		// delete the block meta
+		if err := batch.Delete(iter.Key()); err != nil {
+			return 0, fmt.Errorf("pruning error at height %d: %w", decodeHeightFromKey(iter.Key()), err)
 		}
-		if err := batch.Delete(calcBlockHashKey(meta.BlockID.Hash)); err != nil {
-			return 0, err
+
+		// load the block meta so we can also remove the block hash from the db
+		var pbbm = new(tmproto.BlockMeta)
+		bz := iter.Value()
+
+		err = proto.Unmarshal(bz, pbbm)
+		if err != nil {
+			panic(fmt.Errorf("unmarshal to tmproto.BlockMeta: %w", err))
 		}
-		if err := batch.Delete(calcBlockCommitKey(h)); err != nil {
-			return 0, err
+
+		blockMeta, err := types.BlockMetaFromProto(pbbm)
+		if err != nil {
+			panic(fmt.Errorf("error from proto blockMeta: %w", err))
 		}
-		if err := batch.Delete(calcSeenCommitKey(h)); err != nil {
-			return 0, err
+
+		if err := batch.Delete(blockHashKey(blockMeta.BlockID.Hash)); err != nil {
+			return 0, fmt.Errorf("pruning error at height %d: %w", decodeHeightFromKey(iter.Key()), err)
 		}
-		for p := 0; p < int(meta.BlockID.PartSetHeader.Total); p++ {
-			if err := batch.Delete(calcBlockPartKey(h, p)); err != nil {
-				return 0, err
+
+		pruned++
+		// avoid batches growing too large by flushing to database regularly
+		if pruned%1000 == 0 {
+			if err := iter.Error(); err != nil {
+				return pruned, err
+			}
+			if err := iter.Close(); err != nil {
+				return pruned, err
+			}
+
+			err := batch.Write()
+			if err != nil {
+				return pruned, fmt.Errorf("pruning error at height %d: %w", decodeHeightFromKey(iter.Key()), err)
 			}
+			if err := batch.Close(); err != nil {
+				return pruned, err
+			}
+
+			iter, err = bs.db.Iterator(
+				blockMetaKey(1),
+				blockMetaKey(retainHeight),
+			)
+			if err != nil {
+				panic(err)
+			}
+			defer iter.Close()
+
+			batch = bs.db.NewBatch()
+			defer batch.Close()
+		} else {
+			iter.Next()
 		}
+	}
+	if err := iter.Error(); err != nil {
+		return pruned, err
+	}
+
+	err = batch.WriteSync()
+	if err != nil {
+		return pruned, fmt.Errorf("pruning error at height %d: %w", decodeHeightFromKey(iter.Key()), err)
+	}
+
+	return pruned, nil
+}
+
+// batchDelete is a generic function for deleting a range of values based on the lowest
+// height up to but excluding retainHeight
+func (bs *BlockStore) batchDelete(key func(int64) []byte, retainHeight int64) error {
```

Comment:
I think it's cleaner for this to take the start and end keys instead of passing in a callback function. That would also work across different key structures.